File: gpt-4-turbo/stage2/acc_delete.f90
!T1:acc delete,V:2.7-2.3
      LOGICAL FUNCTION test1()
        USE OPENACC
        IMPLICIT NONE
        INCLUDE "acc_testsuite.Fh"

        INTEGER, PARAMETER :: n = 1024
        REAL, DIMENSION(n) :: a, a_host
        INTEGER :: i, errors = 0

        ! Initialize host array
        DO i = 1, n
          a(i) = REAL(i)
        END DO

        ! Allocate and initialize device array
        !$acc enter data copyin(a(1:n))

        ! Modify array on the device
        !$acc parallel loop present(a)
        DO i = 1, n
          a(i) = a(i) + 1.0
        END DO

        ! Copyout array to host and delete from device
        !$acc exit data copyout(a(1:n))

        ! Validate modification was successful
        errors = 0
        DO i = 1, n
          IF (a(i) /= REAL(i) + 1.0) THEN
            errors = errors + 1
          END IF
        END DO

        ! Use acc_delete to remove the data from the device
        CALL acc_delete(a, SIZEOF(a))

        IF (errors .eq. 0) THEN
          test1 = .FALSE.
        ELSE
          test1 = .TRUE.
        END IF
      END FUNCTION test1

      PROGRAM main
        IMPLICIT NONE
        INTEGER :: failcode, testrun
        LOGICAL :: failed
        INCLUDE "acc_testsuite.Fh"
        LOGICAL :: test1
        failed = .FALSE.
        failcode = 0
        DO testrun = 1, NUM_TEST_CALLS
          failed = failed .or. test1()
        END DO
        IF (failed) THEN
          failcode = failcode + 2 ** 0
          failed = .FALSE.
        END IF
        CALL EXIT(failcode)
      END PROGRAM

File: gpt-4-turbo/stage2/acc_delete.c
#include <stdlib.h>
#include <stdio.h>
#include <openacc.h>

#ifndef SEED
#define SEED 921
#endif

#ifndef NUM_TEST_CALLS
#define NUM_TEST_CALLS 10
#endif

// T1:acc delete, V:2.7-3.3
int test_acc_delete() {
    int err = 0;
    srand(SEED);
    size_t bytes = sizeof(int);
    int *data_host = (int *)malloc(bytes);
    *data_host = rand(); // Fill with random data

    // Allocate and initialize data on the device
    #pragma acc enter data copyin(data_host[0:1])

    // Delete data from device memory
    #pragma acc exit data delete(data_host[0:1])

    // At this point data_host should not be present on the device
    // Trying to perform operation on it without copying back should result in error
    int devicePresent = acc_is_present(data_host, bytes);

    if (devicePresent) {
        printf("Error: data_host should not be present on the device after acc_delete.\n");
        err += 1;
    } else {
        printf("Test passed: data_host is correctly not present on the device after acc_delete.\n");
    }

    free(data_host); // Cleanup host memory

    return err;
}

int main() {
    int failcode = 0;
    int failed = 0;
    for (int x = 0; x < NUM_TEST_CALLS; ++x) {
        failed += test_acc_delete();
    }
    if (failed != 0) {
        failcode += (1 << 0);
    }
    return failcode;
}